### PR TITLE
fix(cachetools): fix decorators signature typing

### LINKED ISSUE #4652: Cachetools stubs are missing types
The new `cachetools` stubs are incomplete and fail in such a case:
```python
    @cachetools.cachedmethod(
        cache=operator.attrgetter("_write_permission_cache"),
        key=functools.partial(cachetools.keys.hashkey, "has_write_permissions"),
    )
```
Returns:

```
error: Argument "cache" to "cachedmethod" has incompatible type "Callable[[Any], Any]"; expected "Cache[partial[Any], <nothing>]"
```
`cache` argument can be a callable as can been seen in the example in the docs: https://cachetools.readthedocs.io/en/stable/#cachetools.cachedmethod

### PR DESCRIPTION
**Pull request by jd**
Fixes #4652

### CODE REVIEW

File: third_party/2and3/cachetools/decorators.pyi [hunk 1]
```diff
@@ -6,6 +6,13 @@ _KT = TypeVar("_KT")
 _VT = TypeVar("_VT")
 _T = TypeVar("_T", bound=Callable[..., Any])
 _T_co = TypeVar("_T_co", covariant=True)
+_T_self = TypeVar("_T_self")
 
-def cached(cache: Cache[_KT, _VT], key: _KT = ..., lock: Optional[ContextManager[_T_co]] = ...) -> Callable[[_T], _T]: ...
-def cachedmethod(cache: Cache[_KT, _VT], key: _KT = ..., lock: Optional[ContextManager[_T_co]] = ...) -> Callable[[_T], _T]: ...
+def cached(
+    cache: Optional[Cache[_KT, _VT]], key: Callable[..., _KT] = ..., lock: Optional[ContextManager[_T_co]] = ...
```
Comment (mariamrf) [inline third_party/2and3/cachetools/decorators.pyi:12]: `cache` here can be any mutable mapping type. From [the documentation](https://cachetools.readthedocs.io/en/stable/#cachetools.cached):

> Note that cache need not be an instance of the cache implementations provided by the cachetools module. cached() will work with any mutable mapping type, including plain dict and weakref.WeakValueDictionary.

So I'm guessing something like this (+ appropriate imports) should work here:

```suggestion
    cache: Optional[MutableMapping[_KT, _VT]], key: Callable[..., _KT] = ..., lock: Optional[ContextManager[_T_co]] = ...
```

I don't know how that would work with the defined stubs for `Cache` though

Comment (jd) [reply third_party/2and3/cachetools/decorators.pyi:12]: That's a good question. Would an `Union` of both be a better choice here or is `MutableMapping` compatible with the definition of `Cache`?

Comment (srittau) [reply third_party/2and3/cachetools/decorators.pyi:12]: `Cache` derives from `cachetools.abc.DefaultMapping`, which in turn derives from `collections.abc.MutableMapping`.

Comment (jd) [reply third_party/2and3/cachetools/decorators.pyi]: Ok, I've updated with `MutableMapping` then!

**Review by srittau — APPROVED** [review]
Thank you for the PR and review.

**Pull request merged**